scripts/protect.sh

```diff
@@ -0,0 +1,235 @@
+}
+
+protectPage() {
+  page="Module:${1}"
+  wiki=$2
+  protectOptions=$3
+  protectMode=$4
+  echo "...wiki = $wiki"
+  echo "...page = $page"
+  wikiApiUrl="${WIKI_BASE_URL}/${wiki}/api.php"
+  ckf="cookie_${wiki}.ck"
+
+  if [[ ${loggedin[${wiki}]} != 1 ]]; then
+    # Login
+    echo "...logging in on \"${wiki}\""
+    loginToken=$(
+      curl \
+        -s \
+        -b "$ckf" \
+        -c "$ckf" \
+        -d "format=json&action=query&meta=tokens&type=login" \
+        -H "User-Agent: ${userAgent}" \
+        -H 'Accept-Encoding: gzip' \
+        -X POST "$wikiApiUrl" \
+        | gunzip \
+        | jq ".query.tokens.logintoken" -r
+    )
+    curl \
+      -s \
+      -b "$ckf" \
+      -c "$ckf" \
+      --data-urlencode "lgname=${WIKI_USER}" \
+      --data-urlencode "lgpassword=${WIKI_PASSWORD}" \
+      --data-urlencode "lgtoken=${loginToken}" \
+      -H "User-Agent: ${userAgent}" \
+      -H 'Accept-Encoding: gzip' \
+      -X POST "${wikiApiUrl}?format=json&action=login" \
+      | gunzip \
+      > /dev/null
+    loggedin[$wiki]=1
+    # Don't get rate limited
+    sleep 4
+  fi
+
+  # Protect Page
+  protectToken=$(
+    curl \
+      -s \
+      -b "$ckf" \
+      -c "$ckf" \
+      -d "format=json&action=query&meta=tokens" \
+      -H "User-Agent: ${userAgent}" \
+      -H 'Accept-Encoding: gzip' \
+      -X POST "$wikiApiUrl" \
+      | gunzip \
+      | jq ".query.tokens.csrftoken" -r
+  )
+  rawProtectResult=$(
+    curl \
+      -s \
+      -b "$ckf" \
+      -c "$ckf" \
+      --data-urlencode "title=${page}" \
+      --data-urlencode "protections=${protectOptions}" \
+      --data-urlencode "reason=Git maintained" \
+      --data-urlencode "expiry=infinite" \
+      --data-urlencode "bot=true" \
+      --data-urlencode "token=${protectToken}" \
+      -H "User-Agent: ${userAgent}" \
+      -H 'Accept-Encoding: gzip' \
+      -X POST "${wikiApiUrl}?format=json&action=protect" \
+      | gunzip
+  )
+  # Don't get rate limited
+  sleep 4
+
+  result=$(echo "$rawProtectResult" | jq ".protect.protections.[].${protectMode}" -r)
+  if [[ $result != *"allow-only-sysop"* ]]; then
+    echo "::warning::could not (${protectMode}) protect $1 on $2"
+    protectErrorMsg="${protectMode}:${wiki}:${page}"
+    protectErrors+=("${protectErrorMsg}")
+  fi
+}
+
+protectExistingPage() {
+  protectPage $1 $2 "edit=allow-only-sysop|move=allow-only-sysop" "edit"
+}
+
+protectNonExistingPage() {
+  protectPage $1 $2 "create=allow-only-sysop" "create"
+}
+
+checkIfPageExists() {
+  page="Module:${1}"
+  wikiApiUrl="${WIKI_BASE_URL}/${2}/api.php"
+  rawResult=$(
+    curl \
+      -s \
+      -b "$ckf" \
+      -c "$ckf" \
+      --data-urlencode "titles=${page}" \
+      --data-urlencode "prop=info" \
+      -H "User-Agent: ${userAgent}" \
+      -H 'Accept-Encoding: gzip' \
+      -X POST "${wikiApiUrl}?format=json&action=query" \
+      | gunzip
+  )
+
+  # Don't get rate limited
+  sleep 4
+
+  if [[ $rawResult == *'missing'* ]]; then
```

Comment:
This fails if the page name includes the word missing I think

Reply:
you think `missing":"` is enough?

Reply:
Probably fine. The clean way would be top use jq, but that would be harder because you need the page id in the path...

Reply:
the problem with jq is that for existing pages it uses pageid (>0) while for missing it goes with "-1", "-2", ...

e.g. for the non existing "Test1234" on commons:
```json
{
    "batchcomplete": "",
    "query": {
        "pages": {
            "-1": {
                "ns": 0,
                "title": "Test1234",
                "missing": ""
            }
        }
    }
}
```
vs the existing "Module:Json":
```json
{
    "batchcomplete": "",
    "query": {
        "pages": {
            "153689": {
                "pageid": 153689,
                "ns": 828,
                "title": "Module:Json"
            }
        }
    }
}
```

Reply:
Well you could check for the page id existing in the path

Reply:
The clean solution would be to use formatversion=2 here, but for now this should suffice.

https://liquipedia.net/commons/api.php?action=query&prop=into&titles=Main_Page|This_Does_Not_Exist&formatversion=2 for an example of formatversion=2